File: packages/typescript/src/features/trust/evaluators/reflection.ts
import { getEntityDetails } from "../../../entities.ts";
import { logger } from "../../../logger.ts";
import {
	type Entity,
	type Evaluator,
	type IAgentRuntime,
	type JsonObject,
	type JsonValue,
	type Memory,
	ModelType,
	type State,
	type UUID,
} from "../../../types/index.ts";
import { composePrompt } from "../../../utils.ts";

const reflectionTemplate = `# Task: Generate Agent Reflection, Extract Facts and Relationships

{{providers}}

# Examples:
{{evaluationExamples}}

# Entities in Room
{{entitiesInRoom}}

# Existing Relationships
{{existingRelationships}}

# Current Context:
Agent Name: {{agentName}}
Room Type: {{roomType}}
Message Sender: {{senderName}} (ID: {{senderId}})

{{recentMessages}}

# Known Facts:
{{knownFacts}}

# Instructions:
1. Generate a self-reflective thought on the conversation about your performance and interaction quality.
2. Extract new facts from the conversation.
3. Identify and describe relationships between entities.
  - The sourceEntityId is the UUID of the entity initiating the interaction.
  - The targetEntityId is the UUID of the entity being interacted with.
  - Relationships are one-direction, so a friendship would be two entity relationships where each entity is both the source and the target of the other.

Generate a response in the following format:
\`\`\`json
{
  "thought": "a self-reflective thought on the conversation",
  "facts": [
      {
          "claim": "factual statement",
          "type": "fact|opinion|status",
          "in_bio": false,
          "already_known": false
      }
  ],
  "relationships": [
      {
          "sourceEntityId": "entity_initiating_interaction",
          "targetEntityId": "entity_being_interacted_with",
          "tags": ["group_interaction|voice_interaction|dm_interaction", "additional_tag1", "additional_tag2"]
      }
  ]
}
\`\`\``;

function resolveEntity(entityId: UUID, entities: Entity[]): UUID {
	if (
		/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(
			entityId,
		)
	) {
		return entityId as UUID;
	}

	let entity: Entity | undefined;

	entity = entities.find((a) => a.id === entityId);
	if (entity?.id) {
		return entity.id;
	}

	entity = entities.find((a) => a.id?.includes(entityId));
	if (entity?.id) {
		return entity.id;
	}

	entity = entities.find((a) =>
		a.names.some((n) => n.toLowerCase().includes(entityId.toLowerCase())),
	);
	if (entity?.id) {
		return entity.id;
	}

	throw new Error(`Could not resolve entityId "${entityId}" to a valid UUID`);
}

async function handler(runtime: IAgentRuntime, message: Memory, state?: State) {
	const { agentId, roomId } = message;

	if (!agentId || !roomId) {
		logger.warn({ message }, "Missing agentId or roomId in message");
		return;
	}

	const [existingRelationships, entities, knownFacts] = await Promise.all([
		runtime.getRelationships({
			entityIds: message.entityId ? [message.entityId] : undefined,
		}),
		getEntityDetails({ runtime, roomId }),
		runtime.getMemories({
			tableName: "facts",
			roomId,
			count: 30,
			unique: true,
		}),
	]);

	const prompt = composePrompt({
		state: {
			...(state?.values || {}),
			knownFacts: formatFacts(knownFacts),
			roomType: message.content.channelType as string,
			entitiesInRoom: JSON.stringify(entities),
			existingRelationships: JSON.stringify(existingRelationships),
			senderId: message.entityId,
		},
		template:
			runtime.character.templates?.reflectionTemplate || reflectionTemplate,
	});

	try {
		const reflection = await runtime.useModel(ModelType.OBJECT_SMALL, {
			prompt,
		});

		if (!reflection) {
			logger.warn({ prompt }, "Getting reflection failed - empty response");
			return;
		}

		if (!reflection.facts || !Array.isArray(reflection.facts)) {
			logger.warn(
				{ reflection },
				"Getting reflection failed - invalid facts structure",
			);
			return;
		}

		if (!reflection.relationships || !Array.isArray(reflection.relationships)) {
			logger.warn(
				{ reflection },
				"Getting reflection failed - invalid relationships structure",
			);
			return;
		}

		const newFacts =
			reflection.facts.filter(
				(fact): fact is JsonObject =>
					fact != null &&
					typeof fact === "object" &&
					!Array.isArray(fact) &&
					!fact.already_known &&
					!fact.in_bio &&
					typeof fact.claim === "string" &&
					fact.claim.trim() !== "",
			) || [];

		await Promise.all(
			newFacts.map(async (fact) => {
				const claim = fact.claim as string;
				const factMemory = await runtime.addEmbeddingToMemory({
					entityId: agentId,
					agentId,
					content: { text: claim },
					roomId,
					createdAt: Date.now(),
				});
				return runtime.createMemory(factMemory, "facts", true);
			}),
		);

		for (const rawRelationship of reflection.relationships) {
			if (
				!rawRelationship ||
				typeof rawRelationship !== "object" ||
				Array.isArray(rawRelationship)
			) {
				continue;
			}

			const relationship = rawRelationship as JsonObject;
			if (
				typeof relationship.sourceEntityId !== "string" ||
				typeof relationship.targetEntityId !== "string" ||
				!Array.isArray(relationship.tags)
			) {
				continue;
			}

			const tags = relationship.tags.filter(
				(t): t is string => typeof t === "string",
			);

			let sourceId: UUID;
			let targetId: UUID;

			try {
				sourceId = resolveEntity(relationship.sourceEntityId as UUID, entities);
				targetId = resolveEntity(relationship.targetEntityId as UUID, entities);
			} catch (error) {
				logger.warn({ error }, "Failed to resolve relationship entities");
				logger.warn({ relationship }, "Unresolved relationship");
				continue;
			}

			const existingRelationship = existingRelationships.find((r) => {
				return r.sourceEntityId === sourceId && r.targetEntityId === targetId;
			});

			if (existingRelationship) {
				const updatedMetadata = {
					...existingRelationship.metadata,
					interactions:
						((existingRelationship.metadata?.interactions as
							| number
							| undefined) || 0) + 1,
				};

				const updatedTags = Array.from(
					new Set([...(existingRelationship.tags || []), ...tags]),
				);

				await runtime.updateRelationship({
					...existingRelationship,
					tags: updatedTags,
					metadata: updatedMetadata,
				});
			} else {
				const metadata =
					relationship.metadata != null &&
					typeof relationship.metadata === "object" &&
					!Array.isArray(relationship.metadata)
						? (relationship.metadata as Record<string, JsonValue>)
						: {};

				await runtime.createRelationship({
					sourceEntityId: sourceId,
					targetEntityId: targetId,
					tags,
					metadata: {
						interactions: 1,
						...metadata,
					},
				});
			}
		}

		await runtime.setCache<string>(
			`${message.roomId}-reflection-last-processed`,
			message?.id || "",
		);

		return {
			success: true,
			text: typeof reflection.thought === "string" ? reflection.thought : "",
		};
	} catch (error) {
		logger.error({ error }, "Error in reflection handler:");
		return;
	}
}

export const reflectionEvaluator: Evaluator = {
	name: "REFLECTION",
	similes: [
		"REFLECT",
		"SELF_REFLECT",
		"EVALUATE_INTERACTION",
		"ASSESS_SITUATION",
	],
	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
	): Promise<boolean> => {
		const lastMessageId = await runtime.getCache<string>(
			`${message.roomId}-reflection-last-processed`,
		);
		const messages = await runtime.getMemories({
			tableName: "messages",
			roomId: message.roomId,
			count: runtime.getConversationLength(),
		});

		if (lastMessageId) {
			const lastMessageIndex = messages.findIndex(
				(msg) => msg.id === lastMessageId,
			);
			if (lastMessageIndex !== -1) {
				messages.splice(0, lastMessageIndex + 1);
			}
		}

		const reflectionInterval = Math.ceil(runtime.getConversationLength() / 4);

		return messages.length > reflectionInterval;
	},
	description:
		"Generate a self-reflective thought on the conversation, then extract facts and relationships between entities in the conversation.",
	handler,
	examples: [
		{
			prompt: `Agent Name: Sarah
Agent Role: Community Manager
Room Type: group
Current Room: general-chat
Message Sender: John (user-123)`,
			messages: [
				{
					name: "John",
					content: { text: "Hey everyone, I'm new here!" },
				},
				{
					name: "Sarah",
					content: { text: "Welcome John! How did you find our community?" },
				},
				{
					name: "John",
					content: { text: "Through a friend who's really into AI" },
				},
			],
			outcome: `{
    "thought": "I'm engaging appropriately with a new community member, maintaining a welcoming and professional tone.",
    "facts": [
        {
            "claim": "John is new to the community",
            "type": "fact",
            "in_bio": false,
            "already_known": false
        },
        {
            "claim": "John found the community through a friend interested in AI",
            "type": "fact",
            "in_bio": false,
            "already_known": false
        }
    ],
    "relationships": [
        {
            "sourceEntityId": "sarah-agent",
            "targetEntityId": "user-123",
            "tags": ["group_interaction"]
        },
        {
            "sourceEntityId": "user-123",
            "targetEntityId": "sarah-agent",
            "tags": ["group_interaction"]
        }
    ]
}`,
		},
	],
};

function formatFacts(facts: Memory[]) {
	return facts
		.reverse()
		.map((fact: Memory) => fact.content.text)
		.join("\n");
}

File: packages/python/elizaos/features/core_capabilities/trust/evaluators/reflection.py
"""Reflection evaluator.

Generates a self-reflective thought on the conversation, then extracts
facts and relationships between entities. Ported from the TypeScript
``reflectionEvaluator``.
"""

from __future__ import annotations

import contextlib
import json
from typing import TYPE_CHECKING, Any

from elizaos.types import Evaluator, EvaluatorResult, HandlerOptions

if TYPE_CHECKING:
    from elizaos.types import ActionResult, IAgentRuntime, Memory, State


REFLECTION_TEMPLATE = """# Task: Generate Agent Reflection, Extract Facts and Relationships

# Entities in Room
{entities_in_room}

# Existing Relationships
{existing_relationships}

# Current Context:
Agent Name: {agent_name}
Room Type: {room_type}
Message Sender: {sender_name} (ID: {sender_id})

# Recent Messages:
{recent_messages}

# Known Facts:
{known_facts}

# Instructions:
1. Generate a self-reflective thought on the conversation about your performance and interaction quality.
2. Extract new facts from the conversation.
3. Identify and describe relationships between entities.
   - The sourceEntityId is the UUID of the entity initiating the interaction.
   - The targetEntityId is the UUID of the entity being interacted with.

Generate a response in the following JSON format:
{{
  "thought": "a self-reflective thought on the conversation",
  "facts": [
      {{
          "claim": "factual statement",
          "type": "fact|opinion|status",
          "in_bio": false,
          "already_known": false
      }}
  ],
  "relationships": [
      {{
          "sourceEntityId": "entity_initiating_interaction",
          "targetEntityId": "entity_being_interacted_with",
          "tags": ["group_interaction", "additional_tag"]
      }}
  ]
}}"""


def _format_facts(facts: list[Any]) -> str:
    """Format fact memories into a text block."""
    lines: list[str] = []
    for fact in reversed(facts):
        content = getattr(fact, "content", None)
        if content:
            text = getattr(content, "text", None) or (
                content.get("text") if isinstance(content, dict) else ""
            )
            if text:
                lines.append(str(text))
    return "\n".join(lines) if lines else "No known facts."


async def _validate(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> bool:
    """Run reflection after enough messages have accumulated since last reflection."""
    if not message.room_id:
        return False

    # Check cache for last processed message
    get_cache = getattr(runtime, "get_cache", None)
    last_message_id = None
    if callable(get_cache):
        cache_key = f"{message.room_id}-reflection-last-processed"
        last_message_id = await get_cache(cache_key)

    # Get recent messages
    get_memories = getattr(runtime, "get_memories", None)
    if not callable(get_memories):
        return False

    get_conv_length = getattr(runtime, "get_conversation_length", None)
    conv_length = get_conv_length() if callable(get_conv_length) else 20
    count = max(conv_length, 10)

    messages = await get_memories(
        table_name="messages",
        room_id=message.room_id,
        count=count,
    )

    if last_message_id:
        # Find the index of the last processed message and only count newer ones
        last_idx = None
        for i, msg in enumerate(messages):
            if getattr(msg, "id", None) == last_message_id:
                last_idx = i
                break
        if last_idx is not None:
            messages = messages[:last_idx]

    reflection_interval = max(1, conv_length // 4)
    return len(messages) > reflection_interval


async def _handler(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
    options: HandlerOptions | None = None,
    responses: list[Memory] | None = None,
    action_results: list[ActionResult] | None = None,
) -> EvaluatorResult:
    """Generate reflection, extract facts and relationships."""
    agent_id = getattr(message, "agent_id", None) or getattr(message, "agentId", None)
    room_id = message.room_id
    entity_id = message.entity_id

    if not agent_id or not room_id:
        return EvaluatorResult.pass_result(score=50, reason="Missing agentId or roomId")

    # Gather context
    get_relationships = getattr(runtime, "get_relationships", None)
    get_memories = getattr(runtime, "get_memories", None)
    get_entities_for_room = getattr(runtime, "get_entities_for_room", None)

    existing_relationships: list[Any] = []
    if callable(get_relationships) and entity_id:
        try:
            existing_relationships = await get_relationships(entity_ids=[entity_id])
        except Exception:
            existing_relationships = []

    entities: list[Any] = []
    if callable(get_entities_for_room):
        try:
            entities = await get_entities_for_room(room_id)
        except Exception:
            entities = []

    known_facts: list[Any] = []
    if callable(get_memories):
        try:
            known_facts = await get_memories(
                table_name="facts",
                room_id=room_id,
                count=30,
                unique=True,
            )
        except Exception:
            known_facts = []

    # Build prompt
    character = getattr(runtime, "character", None)
    agent_name = getattr(character, "name", "Agent") if character else "Agent"
    sender_name = ""
    if state:
        sender_name = getattr(state, "sender_name", "") or getattr(state, "senderName", "")
    channel_type = (
        getattr(message.content, "channel_type", "unknown") if message.content else "unknown"
    )

    prompt = REFLECTION_TEMPLATE.format(
        entities_in_room=json.dumps([str(e) for e in entities] if entities else [], indent=2),
        existing_relationships=json.dumps(
            [str(r) for r in existing_relationships] if existing_relationships else [],
            indent=2,
        ),
        agent_name=agent_name,
        room_type=str(channel_type),
        sender_name=sender_name or str(entity_id),
        sender_id=str(entity_id) if entity_id else "unknown",
        recent_messages=(state.text if state and hasattr(state, "text") and state.text else ""),
        known_facts=_format_facts(known_facts),
    )

    # Call model for reflection
    use_model = getattr(runtime, "use_model", None)
    if not callable(use_model):
        return EvaluatorResult.pass_result(
            score=50, reason="Model API not available for reflection"
        )

    try:
        reflection = await use_model("OBJECT_SMALL", prompt=prompt)
    except Exception:
        return EvaluatorResult.pass_result(score=50, reason="Reflection model call failed")

    if not reflection or not isinstance(reflection, dict):
        return EvaluatorResult.pass_result(
            score=50, reason="Reflection returned empty or invalid response"
        )

    facts = reflection.get("facts", [])
    relationships = reflection.get("relationships", [])

    if not isinstance(facts, list) or not isinstance(relationships, list):
        return EvaluatorResult.pass_result(score=50, reason="Reflection returned invalid structure")

    # Store new facts
    new_facts = [
        f
        for f in facts
        if isinstance(f, dict)
        and not f.get("already_known")
        and not f.get("in_bio")
        and isinstance(f.get("claim"), str)
        and f["claim"].strip()
    ]

    add_embedding = getattr(runtime, "add_embedding_to_memory", None)
    create_memory = getattr(runtime, "create_memory", None)

    if callable(add_embedding) and callable(create_memory):
        for fact in new_facts:
            try:
                fact_memory = await add_embedding(
                    {
                        "entity_id": agent_id,
                        "agent_id": agent_id,
                        "content": {"text": fact["claim"]},
                        "room_id": room_id,
                    }
                )
                await create_memory(fact_memory, "facts", True)
            except Exception:
                pass

    # Process relationships
    update_relationship = getattr(runtime, "update_relationship", None)
    create_relationship = getattr(runtime, "create_relationship", None)

    if callable(create_relationship):
        for raw_rel in relationships:
            if not isinstance(raw_rel, dict):
                continue
            source_id = raw_rel.get("sourceEntityId")
            target_id = raw_rel.get("targetEntityId")
            tags = raw_rel.get("tags", [])
            if (
                not isinstance(source_id, str)
                or not isinstance(target_id, str)
                or not isinstance(tags, list)
            ):
                continue

            tags = [t for t in tags if isinstance(t, str)]

            # Check if relationship already exists
            existing = None
            for rel in existing_relationships:
                rel_source = getattr(rel, "source_entity_id", None) or getattr(
                    rel, "sourceEntityId", None
                )
                rel_target = getattr(rel, "target_entity_id", None) or getattr(
                    rel, "targetEntityId", None
                )
                if str(rel_source) == source_id and str(rel_target) == target_id:
                    existing = rel
                    break

            if existing and callable(update_relationship):
                existing_meta = getattr(existing, "metadata", {}) or {}
                interactions = (existing_meta.get("interactions") or 0) + 1
                existing_tags = getattr(existing, "tags", []) or []
                merged_tags = list(set(list(existing_tags) + tags))
                try:
                    base = vars(existing) if hasattr(existing, "__dict__") else {}
                    update_data = {
                        **base,
                        "tags": merged_tags,
                        "metadata": {**existing_meta, "interactions": interactions},
                    }
                    await update_relationship(update_data)
                except Exception:
                    pass
            else:
                with contextlib.suppress(Exception):
                    await create_relationship(
                        {
                            "sourceEntityId": source_id,
                            "targetEntityId": target_id,
                            "tags": tags,
                            "metadata": {"interactions": 1},
                        }
                    )

    # Update cache with last processed message
    set_cache = getattr(runtime, "set_cache", None)
    if callable(set_cache) and message.id:
        with contextlib.suppress(Exception):
            await set_cache(
                f"{room_id}-reflection-last-processed",
                str(message.id),
            )

    thought = reflection.get("thought", "")
    return EvaluatorResult.pass_result(
        score=75,
        reason=f"Reflection complete: {len(new_facts)} new facts, "
        f"{len(relationships)} relationships processed. "
        f"Thought: {thought[:100] if thought else '(none)'}",
    )


reflection_evaluator = Evaluator(
    name="REFLECTION",
    description=(
        "Generate a self-reflective thought on the conversation, then "
        "extract facts and relationships between entities."
    ),
    handler=_handler,
    validate=_validate,
)
